src/main/kotlin/com/openlattice/shuttle/IntegrationService.kt

```diff
@@ -0,0 +1,349 @@
+                    }
+                }
+            }
+        }
+    }
+
+    fun enqueueIntegrationJob(integrationName: String, integrationKey: UUID): UUID {
+        checkIntegrationExists(integrationName)
+        val integration = integrations.getValue(integrationName)
+        checkState(integrationKey == integration.key, "Integration key $integrationKey is incorrect")
+        val jobId = generateIntegrationJobId()
+        jobQueue.put(jobId)
+        integrationJobs[jobId] = IntegrationJob(integrationName, IntegrationStatus.QUEUED)
+        return jobId
+    }
+
+    private final fun loadCargo(jobId: UUID) {
+        val integrationJob = integrationJobs.getValue(jobId)
+        val integration = integrations.getValue(integrationJob.integrationName)
+        val token = getIdToken(missionParameters.auth)
+        val apiClient = ApiClient(integration.environment) { token }
+        val dataIntegrationApi = apiClient.dataIntegrationApi
+
+        //an integration object is expected to have a non-empty logEntitySetId,
+        //a non-null flight, and non-null key in order for the integration to run successfully
+        val flightPlan = mutableMapOf<Flight, Payload>()
+        val tableColsToPrint = mutableMapOf<Flight, List<String>>()
+        integration.flightPlanParameters.values.forEach {
+            val srcDataSourceProperties = Properties()
+            srcDataSourceProperties.putAll(it.source)
+            val srcDataSource = getSrcDataSource(srcDataSourceProperties)
+            val payload = JdbcPayload(readRateLimit.toDouble(), srcDataSource, it.sql, fetchSize, readRateLimit != 0)
+            flightPlan[it.flight!!] = payload
+            tableColsToPrint[it.flight!!] = it.sourcePrimaryKeyColumns
+        }
+        val destinationsMap = generateDestinationsMap(integration, missionParameters, dataIntegrationApi)
+
+        val shuttle = Shuttle(
+                integration.environment,
+                flightPlan,
+                entitySets.values.associateBy { it.name },
+                entityTypes.values.associateBy { it.id },
+                propertyTypes.values.associateBy { it.type },
+                destinationsMap,
+                dataIntegrationApi,
+                tableColsToPrint,
+                missionParameters,
+                StorageDestination.S3,
+                blackbox,
+                Optional.of(entitySets.getValue(integration.logEntitySetId.get())),
+                Optional.of(jobId),
+                idService,
+                hazelcastInstance
+        )
+
+        executor.submit {
+            semaphore.acquire()
+            shuttle.launch(uploadBatchSize)
+        }.addListener(Runnable {
+            integration.callbackUrls.ifPresent {
+                submitCallback(jobId, it)
+            }
+            semaphore.release()
+        }, executor)
+    }
+
+    fun pollIntegrationStatus(jobId: UUID): IntegrationStatus {
+        checkIntegrationJobExists(jobId)
```

Comment:
Since this does an additional lookup on a hazelcast map which is expensive, you might want to build this into the nullable case, i.e.
```
return integrationJobs.get(jobId)?.integrationStatus ?: throw IllegalStateException("Job Id $jobId is not assigned to an existing integration job")
```